fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MultidimensionalIndexScanBounds.java

```diff
@@ -0,0 +1,314 @@
+        };
+
+        boolean overlapsMbr(@Nonnull RTree.Rectangle mbr);
+
+        boolean containsPosition(@Nonnull RTree.Point position);
+    }
+
+    /**
+     * Scan bounds that consists of other {@link SpatialPredicate}s to form a logical OR.
+     */
+    public static class Or implements SpatialPredicate {
+        @Nonnull
+        private final List<SpatialPredicate> children;
+
+        public Or(@Nonnull final List<SpatialPredicate> children) {
+            this.children = ImmutableList.copyOf(children);
+        }
+
+        @Override
+        public boolean overlapsMbr(@Nonnull final RTree.Rectangle mbr) {
+            return children.stream()
+                    .anyMatch(child -> child.overlapsMbr(mbr));
+        }
+
+        @Override
+        public boolean containsPosition(@Nonnull final RTree.Point position) {
+            return children.stream()
+                    .anyMatch(child -> child.containsPosition(position));
+        }
+
+        @Override
+        public String toString() {
+            return children.stream().map(Object::toString).collect(Collectors.joining(" or "));
+        }
+    }
+
+    /**
+     * Scan bounds that consists of other {@link SpatialPredicate}s to form a logical AND.
+     */
+    public static class And implements SpatialPredicate {
+        @Nonnull
+        private final List<SpatialPredicate> children;
+
+        public And(@Nonnull final List<SpatialPredicate> children) {
+            this.children = ImmutableList.copyOf(children);
+        }
+
+        @Override
+        public boolean overlapsMbr(@Nonnull final RTree.Rectangle mbr) {
+            return children.stream()
+                    .allMatch(child -> child.overlapsMbr(mbr));
+        }
+
+        @Override
+        public boolean containsPosition(@Nonnull final RTree.Point position) {
+            return children.stream()
+                    .allMatch(child -> child.containsPosition(position));
+        }
+
+        @Override
+        public String toString() {
+            return children.stream().map(Object::toString).collect(Collectors.joining(" and "));
+        }
+    }
+
+    /**
+     * Scan bounds describing an n-dimensional hypercube.
+     */
+    public static class Hypercube implements SpatialPredicate {
+        @Nonnull
+        private final List<TupleRange> dimensionRanges;
+
+        public Hypercube(@Nonnull final List<TupleRange> dimensionRanges) {
+            this.dimensionRanges = ImmutableList.copyOf(dimensionRanges);
+        }
+
+        @Override
+        public boolean overlapsMbr(@Nonnull final RTree.Rectangle mbr) {
+            Preconditions.checkArgument(mbr.getNumDimensions() == dimensionRanges.size());
+
+            for (int d = 0; d < mbr.getNumDimensions(); d++) {
+                final Tuple lowTuple = Tuple.from(mbr.getLow(d));
+                final Tuple highTuple = Tuple.from(mbr.getHigh(d));
+
+                final TupleRange dimensionRange = dimensionRanges.get(d);
```

Comment:
I believe the following 1-D overlap predicates could be methods on `TupleRange` itself.

Reply:
done.